### PR TITLE
[DT-138] PHPStan: gain the official extension status

### PR DESCRIPTION
**Pull request by pushokwhite**
#### Overview

- Ticket: https://spryker.atlassian.net/browse/DT-138

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+language: php
+php:
+  - 7.1
+  - 7.3
+before_script:
+  - composer self-update
```
Comment (dereuromark) [inline .travis.yml]: not needed

File: README.md [hunk 2]
```diff
@@ -3,48 +3,24 @@
 [![License](https://poser.pugx.org/spryker/code-sniffer/license.svg)](https://packagist.org/packages/spryker-sdk/phpstan-spryker)
 
 
-## Documentation
-See https://github.com/phpstan/phpstan
+* [PHPStan](https://github.com/phpstan/phpstan)
+
+This extension provides following features:
+
+* Notifies you when you try to use `getConfig`, `getClient`, `getFacade`, `getFactory`, `getSharedConfig`, `getQueryContainer` methods without specifying theirs types in a class's DocBlock.
```
Comment (dereuromark) [inline README.md]: () per method

**Comment by dereuromark** [comment]
Looks good.
We we will also need feedback from @ondrejmirtes here then.

File: .travis.yml (hunk 1 above)
Comment (pushokwhite) [reply .travis.yml]: https://docs.travis-ci.com/user/languages/php/#installing-composer-packages
Correct, line was deleted.

File: phpstan.neon [hunk 3]
```diff
@@ -0,0 +1,11 @@
+includes:
+	- vendor/phpstan/phpstan-strict-rules/rules.neon
```
Comment (ondrejmirtes) [inline phpstan.neon]: This should also include `extension.neon`.

File: src/Type/Spryker/ClassDynamicTypeExtension.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+			throw new \PHPStan\ShouldNotHappenException();
+		}
+
+		$docComment = $scope->getClassReflection()->getNativeReflection()->getDocComment();
+
+		if ($docComment === false) {
+			throw new \Exception('Please add PHPDoc block with @method annotation for "getFactory(), getQueryContainer() and/or getFacade()" if one  is used.');
+		}
+
+		preg_match_all('#@method\s+(?:(?P<IsStatic>static)\s+)?(?:(?P<Type>[^\(\*]+?)(?<!\|)\s+)?(?P<MethodName>[a-zA-Z0-9_]+)(?P<Parameters>(?:\([^\)]*\))?)#', $docComment, $matches, PREG_SET_ORDER);
```
Comment (ondrejmirtes) [inline src/Type/Spryker/ClassDynamicTypeExtension.php]: This should make use of `AnnotationsMethodsClassReflectionExtension` or at least get inspired by the code in it. Parsing phpDocs with regexes is obsolete.

File: src/Type/Spryker/ClassDynamicTypeExtension.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		}
+
+		if (!$methodCall->name instanceof Identifier) {
+			throw new \PHPStan\ShouldNotHappenException();
+		}
+
+		$docComment = $scope->getClassReflection()->getNativeReflection()->getDocComment();
+
+		if ($docComment === false) {
+			throw new \Exception('Please add PHPDoc block with @method annotation for "getFactory(), getQueryContainer() and/or getFacade()" if one  is used.');
```
Comment (ondrejmirtes) [inline src/Type/Spryker/ClassDynamicTypeExtension.php]: Extensions should never throw exceptions, unless it's a completely inexpected situations. If you're tempted to throw exceptions from an extension, you should write a corresponding rule instead (and return `MixedType` or `ErrorType` here).

File: phpstan.neon (hunk 3 above)
Comment (pushokwhite) [reply phpstan.neon]: Should it? I mean - this file is only for build testing, I could move it to `tests/` folder to clarify its purpose.

**Comment by pushokwhite** [comment]
@ondrejmirtes, changes were made. Please look once you have some free time.

**Comment by dereuromark** [comment]
ping @ondrejmirtes Are you OK with the current changes?

**Comment by dereuromark** [comment]
@ondrejmirtes We are a bit blocked here. If you are not available or not OK with the current code, we will have to put this task on hold then. No problem.

**Comment by dereuromark** [comment]
Seems like this is now cancelled then. We will use this repo from now on instead directly.

**Pull request merged**